### ISSUE TITLE
FEATURE: Save and load object X-ray states in BCF 

### ISSUE THREAD

**Issue by xeolabs** [xeokit/xeokit-sdk#1247]
This [PR](https://github.com/xeokit/xeokit-sdk/pull/1257) adds to `BCFViewpointsPlugin` a new `xrayAsZeroAlpha` option, which specifies how the plugin should handle X-ray effect for BCF viewpoints.

> Published in  xeokit-sdk 2.4.2-beta-10 and  xeokit-bim-viewer 2.4.16

````javascript
const bcfViewpoints = new BCFViewpointsPlugin(viewer, {
    xrayAsZeroAlpha: true // Default is false
});
````

When `xrayAsZeroAlpha` is `true`:

* when saving a viewpoint, for each **X-rayed object**, save `0` in the object's BCF color opacity 
* when saving a viewpoint, for each **non-X-rayed object**, save the object's colorized opacity in its BCF color opacity, as usual 
* when loading a viewpoint saved from a xeokit application (i.e. BCF `originating_system == "xeokit.io"`), if an object's BCF color opacity is `0` then set that object X-rayed
* when loading a viewpoint saved from a xeokit application, if an object's BCF color opacity is not `0`, then set the opacity of that object to that color opacity, as usual.

The semantic here is that when `xrayAsZeroAlpha` is enabled, a saved BCF viewpoint sets color opacity `0` to indicate the object is X-rayed. When loading, the BCF color opacity `0` means the object is X-rayed. This assumes that we never colorize an object to `0` opacity when we use `xrayAsZeroAlpha`. That would make the object invisible, and that can be better done by simply setting it invisible anyway.

When `xrayAsZeroAlpha` is `false` (default):

* don't save any X-ray information in the `A` channel of each `coloring` component 
* reset all objects as not X-rayed whenever loading a viewpoint.
 
## Rationale

* Ensures that color opacities in BCF saved from other BIM applications are treated as colorize opacities, and not as X-ray, because X-ray is not supported in BCF and that would break assumption made by those applications.
* Ensures that no extra fields (i.e. `xray`) are added to BCF viewpoints, which would break compatibility with the BCF schema.
* If a xeokit application will save and load the same BCF, it must be careful not to colorize any objects with opacity == '0' 

## Demos

* [Saving BCF with xrayAsZeroAlpha](https://xeokit.github.io/xeokit-sdk/examples/bookmarking/#saveViewpoint_xrayAsZeroAlpha)
* [Loading BCF with xrayAsZeroAlpha](https://xeokit.github.io/xeokit-sdk/examples/bookmarking/#loadViewpoint_xrayAsZeroAlpha)

## What this solves

Users have found that BCF viewpoints don't properly capture and restore xeokit's "x-ray objects" effect.

* xeokit-sdk has the concept of "x-raying" objects, where in the Viewer we can apply the X-ray effect to objects and mark them as X-rayed, ie. e can enable or disable "x-ray" for each object 
* However, the BCF specification does not have the concept of "x-raying". When we save a BCF viewpoint, all the viewpoint can do is save RGBA color values with A as a value less than 1. The BCF doesn't indicate that A<1 could mean x-ray.
* Then when loading a BCF viewpoint back into a Viewer, or another Viewer, the Viewer doesn't know that A<1 could mean that the object is X-rayed.

**Comment by Amoki** [comment]
Be careful, BCF is ARGB, not RGBA.

This is something I encountered many years ago (https://github.com/xeokit/xeokit-sdk/blob/master/src/plugins/BCFViewpointsPlugin/BCFViewpointsPlugin.js#L511)

I'm not sure about the custom ` xrayedObjects` field that needs explicit compatibility, but I think we can use the `originating_system` and/or `authoring_tool_id` fields (either on Viewpoint or on Components)

There is no reason for A to be equal to 0 (we use the visibility for that) but I see many reasons for A to be < 1 and not wanting to have x-rayed stuff.

If `originating_system` is xeokit and A == 0 then x-rayed

**Comment by xeolabs** [comment]
>> If originating_system is xeokit and A == 0 then x-rayed

Yes that seems like a good workaround. 

I will also add something like a `xrayAsZeroApha` option, to the `BCFViewpointsPlugin` to use when saving and load:

* when saving, and `xrayAsZeroAlpha` is `true`, set A == 0 for each X-rayed object
* when loading with `originating_system` is xeokit and `xrayAsZeroAlpha` is `true`, set object X-rayed when A == 0
* `xrayAsZeroAlpha` is `false` by default (or should it be `true` by default?)

**Comment by xeolabs** [comment]
@Amoki would it be so bad if xeokit's BCF viewpoints containing an additional `xrayed` components field, even if that's not part of the BCF spec?

Then other BIM apps could just ignore the field, but xeokit would be able to use it. 

Using the `originating_system` and this extra field, we could allow BCF colors to zero opacities.

**Comment by Amoki** [comment]
> would it be so bad if xeokit's BCF viewpoints containing an additional xrayed components field

No. If the goal is to retrieve jsons from `getViewpoint()` and replay them with `setViewpoint()`, it works well. If the goal is to use a [BCF-API](https://github.com/buildingSMART/BCF-API), it won't work without modifications to the API implementation that may not be possible in some cases.

**Comment by Amoki** [comment]
Next version of BCF adds what we want here: https://github.com/buildingSMART/BCF-XML/pull/363/files

**Comment by xeolabs** [comment]
@Amoki I just went ahead and supported X-ray using the new BCF translucency compenents: https://github.com/xeokit/xeokit-sdk/pull/1278

Published in @xeokit/xeokit-sdk 2.4.2-beta-30
